Migrate the RECEIVE_PLUGINS error notice and remove lib/sites-list

jsnajdr opened the pull request:
This PR removes the last bit of code that uses anything from `lib/sites-list`. Namely, showing a "compressed" error notice when fetching site plugins fails:

<img width="345" alt="screen shot 2018-04-16 at 22 49 14" src="https://user-images.githubusercontent.com/664258/38862378-0a39b1c2-4235-11e8-9678-ee13d4d91a0d.png">

The rest is a series of janitorial commits that remove the `lib/sites-list` code and most (not all) references from comments and docs.

The impact on bundle size is not bad:
```
chunk                            stat_size           parsed_size           gzip_size
async-load-design~plugins~stats     +313 B  (+0.5%)       +250 B  (+0.7%)      +65 B  (+1.0%)
devdocs                              -80 B  (-0.0%)        -80 B  (-0.0%)      -13 B  (-0.0%)
manifest                              +0 B                  +0 B                -3 B  (-0.1%)
build                              -9286 B  (-0.3%)      -4681 B  (-0.4%)    -1164 B  (-0.4%)
```

**How to test:**
1. Have a few Jetpack/Atomic sites.
2. Go to `/plugins` (all sites selected, shows a list of plugins from WP.org)
3. In the Network devtool, toggle offline mode so that REST requests will start fail
4. Click on "Manage"
5. Check that all requests to fetch site plugins failed and a compressed notice (Error fetching plugins on N sites) is shown.

Fixes #24144.

jsnajdr committed:
Plugin Notices: guard for undefined 'plugin' in log record

Don't crash if log record doesn't have `plugin` property: guard property
accesses with `_.get`.

There's a little fix of conditional in `getSuccessHref`: `&&` to `||`.

Also improves the computation of `numberOfSites` and `numberOfPlugins` by
using `_.uniqBy`.

jsnajdr committed:
Plugin Notices: add support for RECEIVE_PLUGINS error logging

Adds support for `RECEIVE_PLUGINS` action to Plugins Log Store and stops
mounting the `SitesListNotices` component in the main layout -- it doesn't
do anything useful no more.

jsnajdr committed:
Remove the lib/sites-list modules

The last use case (RECEIVE_PLUGINS notice) has just been migrated to another place.

jsnajdr committed:
Sites Reducer: remove unneccessary site object validation

Remove a `_.pick` call that filters out unwanted properties from a site object.
That was useful for bridging between legacy `sites-list` and Redux. But today,
all site objects come from REST API responses only and are sufficiently validated
with a JSON schema.

jsnajdr committed:
Remove references to lib/sites-list from docs and comments

jsnajdr committed:
Remove lib/sites-list from list of prohibited modules in ESLint config

The `lib/sites-list` doesn't exist any more, so guarding the import can
be removed.

a8ci18n commented on client/lib/plugins/notices.jsx:
```diff
@@ -970,6 +969,11 @@ export default {
 				}
+
+			case 'RECEIVE_PLUGINS':
+				return i18n.translate( 'Error fetching plugins on %(site)s.', {
```
Hi! I've found a possible matching string that has already been translated **26** times:
`translate( 'Error fetching plugins on %(siteName)s.' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

ockham commented:
CircleCI chokes because of two failing unit tests in `client/state/sites/test/reducer` :slightly_frowning_face:

ockham commented on client/blocks/site/README.md:
```diff
@@ -1,7 +1,7 @@
 ==========
 
-This component displays a Site item using the data structure provided by sites-list. It's used to display context and to render the picker.
+This component displays a Site item using the data structure provided by Redux sites list. It's used to display context and to render the picker.
```
Maybe reword a little more, e.g.

> This component displays a Site item using data from the Redux `sites` subtree.

jsnajdr committed:
Remove obsolete tests for ignoring extra properties when receiving sites

That functionality was relevant when receiving `sites-list` items, which were objects
with methods and computed properties. Now all the sites are REST API responses.

jsnajdr committed:
Improve the docs of blocks/site a little more

jsnajdr commented:
Thanks @ockham for looking at this. I "fixed" the two failing tests by deleting them and improved the README of `blocks/site` a little. I believe the PR is ready for further review now.

gwwar commented on client/components/plans/plan-price/index.jsx:
```diff
@@ -19,7 +19,7 @@ class PlanPrice extends React.Component {
 
 		if ( plan ) {
-			// the properties of a plan object from sites-list is snake_case
+			// the properties of a plan object from sites list is snake_case
```
Do we still have snake_case properties?

jsnajdr replied:
@gwwar I don't know at all what the story of snake_case properties is. Is it something legacy that we can safely remove?

gwwar replied:
From the comment, it sounded like sites-list was not normalizing properties like what we were doing in the redux store. So theoretically we can remove this entire comment.

jsnajdr replied:
So I did a little research and learned that:
- site plans (`state.sites.plans`, the `getSitePlan` selector) have camelCase properties, because the API responses are transformed (search for `createSitePlanObject` function)
- global plans (`state.plans`, the `getPlanBySlug` selector) have snake_case properties, because the API responses are not transformed when receiving them

I updated the comment accordingly.

This whole component (`components/plans/plan-price`) will be removed soon anyway, because it's used only by `PremiumPopover`, used in legacy (non-Kracken) domains search. @jsnmoon has been landing relevant PRs in last few days.

gwwar commented:
We may need a create a Desktop PR on a branch with the same name as this one, it's currently failing at. If we remove/add needs review label the desktop/canary tests should re-run
```
ERROR in ./calypso/server/boot/index.js
Module build failed: Error: Plugin/Preset files are not allowed to export objects, only functions. In /Users/distiller/wp-desktop/calypso/node_modules/babel-preset-stage-2/lib/index.js
    at createDescriptor (/Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-descriptors.js:181:11)
    at /Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-descriptors.js:106:12
    at Array.map (<anonymous>)
    at createDescriptors (/Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-descriptors.js:105:27)
    at createPresetDescriptors (/Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-descriptors.js:97:10)
    at /Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-descriptors.js:69:19
    at presets (/Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-descriptors.js:59:25)
    at mergeChainOpts (/Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-chain.js:324:68)
    at /Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-chain.js:277:7
    at buildRootChain (/Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/config-chain.js:111:20)
    at loadPrivatePartialConfig (/Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/partial.js:51:53)
    at loadFullConfig (/Users/distiller/wp-desktop/node_modules/@babel/core/lib/config/full.js:43:37)
    at /Users/distiller/wp-desktop/node_modules/@babel/core/lib/transform.js:28:33
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
 @ ./desktop/server/server.js 31:13-28
 @ ./desktop/server/index.js
 @ ./desktop/app.js
 @ ./desktop/index.js
```

gwwar commented:
Awesome work 💖 Code changes look good and this smoke tests well for me. Let's get the Desktop/e2e tests green before 🚢 

<img width="390" alt="screen shot 2018-04-17 at 5 11 57 pm" src="https://user-images.githubusercontent.com/1270189/38905003-941b32c6-4262-11e8-9448-5d8af6f228e1.png">

akirk commented on client/lib/plugins/notices.jsx:
```diff
@@ -763,6 +758,10 @@ export default {
 				}
 				break;
+			case 'RECEIVE_PLUGINS':
+				return this.props.translate( 'Error fetching plugins on %(numberOfSites)d sites.', {
```
Please add plural handling here. Thanks!

jsnajdr replied:
This line is executed only when `numberOfSites > 1`. The singular case is handled on line 974, in the `singleErrorMessage` function.

jsnajdr replied:
@akirk For future refactorings, is it possible to merge both cases into one call, isn't it?
```js
translate(
  'Error fetching plugins on %(site)s',
  'Error fetching plugins on %(siteCount)d',
  {
    count: sites.length,
    args: {
      siteCount: sites.length,
      site: sites[0].name,
	}
  }
);
```

akirk replied:
Please don't do this. Japanese for example doesn't have plurals, so it will always use the first case, even if you have a count > 1.

akirk replied:
Also, you cannot keep it the way it is because it cannot handle different plural phrases for other languages. For example Russian has different plurals for (2, 3 and 4) and (5, 6).

Too late, it's merged.

jsnajdr replied:
Thanks for pointing these subtle bugs out, @akirk! This plurals bug in log-store notices was there forever since Calypso was first released, and in this PR, I just moved some code to a different location, without really changing it.

It' hard to fix without an almost complete rewrite. Which will fortunately happen soon as we convert the `log-store` stores to Redux.

akirk replied:
Alright, no problem. I just caught sight of it here and it's a fair point that you shouldn't be blamed for errors in code you moved :) All good!

jsnajdr commented:
@gwwar I think the wp-desktop build fails because of Babel 7: my PR was based on a master commit that's still on Babel 6, while wp-desktop/master is Babel 7. #24284 is failing with exactly the same error. I rebased them both and triggered a rebuild.

jsnajdr committed:
Improve a comment about plan properties

jsnajdr commented:
@ockham @gwwar I believe all issues in this PR are resolved, including the mysteriously failing wp-desktop and e2e checks. If I get an approval, I'll merge later today.

sirreal reviewed (approved):
I haven't looked deeply at all the code, but it's received several reviews.

I've smoke tested and full e2e suite is all clear. I think this is safe to merge.

Nice improvement 🎉
